Updated supported versions for Ubuntu: 12.04, 14.04, 16.04, and 16.10 (best effort)

rallytime opened the pull request:
The bootstrap script should match the supported versions of SaltStack
more closely, and we don't support versions that are considered EOL by
the OS distributors themselves.

This change updates the bootstrap script to support Ubuntu 12.04, 14.04,
16.04, and 16.10 (which the .10 release is a "best-effort" support since
it is a non-LTS OS release from Ubuntu).

@vutny I'd appreciate a review on this one as well if you have a moment.

Commit:
Updated supported versions for Ubuntu: 12.04, 14.04, 16.04, and 16.10 (best effort)

The bootstrap script should match the supported versions of SaltStack
more closely, and we don't support versions that are considered EOL by
the OS distributors themselves.

This change updates the bootstrap script to support Ubuntu 12.04, 14.04,
16.04, and 16.10 (which the .10 release is a "best-effort" support since
it is a non-LTS OS release from Ubuntu).

Commit:
Remove linuxmint derivative mappings for EOL ubuntu OSes

Commit:
Remove Ubuntu 15 from ubuntu_codename_translation function

rallytime commented:
Refs #972

vutny commented on bootstrap-salt.sh:
```diff
@@ -1201,12 +1201,6 @@ __ubuntu_derivatives_translation() {
-    # https://bugs.launchpad.net/linuxmint/+bug/1198751
-
-    linuxmint_16_ubuntu_base="13.10"
     linuxmint_17_ubuntu_base="14.04"
```
Since we are dropping some of Linux Mint releases support, shouldn't we also reflect this in the `README` file?

rallytime replied:
Yeah I looked around for any documented support for Linux Mint and didn't see any, but I should probably add some.

EDIT: Just kidding. There is Linux Mint support listed in the README. IDK why I didn't find it before. I'll adjust that.

vutny commented on bootstrap-salt.sh:
```diff
@@ -2303,13 +2295,8 @@ __enable_universe_repository() {
     # Ubuntu versions higher than 12.04 do not live in the old repositories
-    if [ "$DISTRO_MAJOR_VERSION" -gt 12 ] || ([ "$DISTRO_MAJOR_VERSION" -eq 12 ] && [ "$DISTRO_MINOR_VERSION" -gt 04 ]); then
+    if [ "$DISTRO_MAJOR_VERSION" -gt 12 ]; then
         add-apt-repository -y "deb http://archive.ubuntu.com/ubuntu $(lsb_release -sc) universe" || return 1
```
And since this PR drops support for Ubuntu major releases before the `12.04`,  the above `if` statement is totally unnecessary.
The `precise` release is still in the main archive: http://archive.ubuntu.com/ubuntu/dists/precise/universe/.

rallytime replied:
Great! That will be much cleaner! :D

vutny commented on bootstrap-salt.sh:
```diff
@@ -2303,13 +2295,8 @@ __enable_universe_repository() {
-        add-apt-repository -y "deb http://old-releases.ubuntu.com/ubuntu $(lsb_release -sc) universe" || return 1
     fi
 
     add-apt-repository -y "deb http://old-releases.ubuntu.com/ubuntu $(lsb_release -sc) universe" || return 1
```
That line need to be removed, there are no LTS releases hosted on http://old-releases.ubuntu.com/ubuntu

rallytime replied:
Ok, can do. I was wondering about why that was there, but didn't want to remove it if there was a reason.

vutny commented on bootstrap-salt.sh:
```diff
@@ -2385,7 +2364,7 @@ install_ubuntu_deps() {
 
 install_ubuntu_stable_deps() {
-    if ([ "${_SLEEP}" -eq "${__DEFAULT_SLEEP}" ] && [ "$DISTRO_MAJOR_VERSION" -lt 15 ]); then
+    if ([ "${_SLEEP}" -eq "${__DEFAULT_SLEEP}" ] && [ "$DISTRO_MAJOR_VERSION" -lt 16 ]); then
```
I guess the parentheses could be safely omitted here... :smiley:

vutny commented on bootstrap-salt.sh:
```diff
@@ -2476,24 +2450,12 @@ install_ubuntu_stable_deps() {
-        __apt_get_install_noinput software-properties-common || return 1
-    else
-        __apt_get_install_noinput python-software-properties || return 1
-    fi
```
This code here is for a reason, just the expression and comment are little bit wrong.
On Ubuntu 12.04 you need to install `python-software-properties` to get `add-apt-repository` binary:
http://packages.ubuntu.com/precise/all/python-software-properties/filelist

It is not in the `software-properties-common` pkg though: http://packages.ubuntu.com/precise/all/software-properties-common/filelist

The `if` statement should be:
```sh
if [ "$DISTRO_MAJOR_VERSION" -gt 12 ]; then
    ...
```

rallytime replied:
Ah, so this logic wasn't quite right to begin with. Interesting...I'll fix that.

vutny commented on bootstrap-salt.sh:
```diff
@@ -2599,7 +2561,7 @@ install_ubuntu_git() {
-    if [ "$DISTRO_MAJOR_VERSION" -lt 15 ]; then
+    if [ "$DISTRO_MAJOR_VERSION" -lt 16 ]; then
         return 0
     fi
```
Looks like this whole `if` statement is really a legacy of the previous versions. We do not trying to start Salt daemons here in the post install function for Ubuntu, but only enabling them to start on boot. Generally it is done automatically for all packages regardless of distro version, but the later code just double checks that.

I think the above statement could be safely removed.

rallytime replied:
Good call.

vutny reviewed (changes requested):
This is really awesome @rallytime !

I think we just need to handle few things more carefully :smiley:

Commit:
Various fixes from PR review:

- Update Linux Mint supported versions in README file
- Remove if statement restricting archive.ubuntu.com repository addition
- Remove old-releases.ubuntu.com repository addition
- Remove some unnecessary parens in an if statement && clause
- Add the python-software-properties install logic back in for Ubutu 12
- Remove the workaround for starting deamons for older versions (no longer applicable)
- Fix architecture spelling in README file

rallytime commented:
@vutny Thanks for taking such a careful look here! I have made all of the changes you suggested. Please take a look and make sure I didn't miss anything.

vutny commented on bootstrap-salt.sh:
```diff
@@ -2430,10 +2404,10 @@ install_ubuntu_stable_deps() {
-                UBUNTU_VERSION=14.04
-                UBUNTU_CODENAME=trusty
+                UBUNTU_VERSION=16.04
+                UBUNTU_CODENAME=xenial
```
@rallytime This is what I've mentioned in comments for PR #1001 .

The above and later code explicitly says:
**If** we are on 16.10 "yakkety" **then** go configure the SaltStack's repo for 16.04 "xenial", which is:
```
https://repo.saltstack.com/apt/ubuntu/16.04/amd64/latest/SALTSTACK-GPG-KEY.pub
```

This is the "best effort" approach introduced before for Ubuntu non-LTS releases and now it simply works for 16.10 too!
I have tested it and got latest Salt packages successfully installed (and running) from SaltStack's repo on "yakkety", while all other dependencies were fetched from Ubuntu's main and universe repositories.

I think this is really nice, now we have full Ubuntu 16.10 support! That's worth adding to the README :smiley: 

Anyhow, the changes in #1001 still make sense because for now the packages from "xenial" repo are compatible and installable on the "yakkety" system, but that could be changed any moment.

rallytime replied:
Good call! I'll add a note about that in the README. Thanks!

vutny commented on bootstrap-salt.sh:
```diff
@@ -2430,10 +2404,10 @@ install_ubuntu_stable_deps() {
         if [ "$(echo "$STABLE_REV" | egrep '^(2015\.5|2015\.8|2016\.3|2016\.11|latest|archive\/)')" != "" ]; then
             # Workaround for latest non-LTS ubuntu
-            if [ "$DISTRO_MAJOR_VERSION" -eq 15 ]; then
+            if [ "$DISTRO_MAJOR_VERSION" -eq 16 ] && [ "$DISTRO_MINOR_VERSION" -eq 10 ]; then
```
I think we can just use this here:
```bash
if [ "$DISTRO_VERSION" = "16.10" ]; then
```

rallytime replied:
Oh, haha. Yes. Let's make that simpler.

vutny reviewed (changes requested):
@rallytime Again, this PR is absolutely awesome. While dropping lots of legacy stuff, it also brings 16.10 support and fixes some long standing issues. Well done!

I have only one minor suggestion... :smiley: Please look at the in-line comments. Thanks!

Commit:
Simplify ubuntu version check to look for "16.10"

Commit:
Add note about best-effort support for non-lts ubuntu releases to README

rallytime commented:
@vutny I made some more updates per your suggestions. Please take another look and let me know what you think.

vutny commented on README.rst:
```diff
@@ -255,6 +255,20 @@ Ubuntu and derivatives
+script will setup the repository for Ubuntu 16.04 from 
+[repo.saltstack.com](http://repo.saltstack.com/#ubuntu) and 
+install the 16.04 packages.
+
```
Maybe it would be better to format links using RsT syntax? Like this:
```
text text `SaltStack corporate repository for Ubuntu`_ some other text

.. _`SaltStack corporate repository for Ubuntu`: http://repo.saltstack.com/#ubuntu
```

This will probably look more clear when the file will be rendered on GitHub.
Also, the line width of 100 chars was used before.

rallytime replied:
OK. I just have been doing too much context switching. I'll fix it.

vutny reviewed (changes requested):
@rallytime, I know I'm annoying, sorry about that :wink: 
Please take a look at the comment I made.

Commit:
Use rst syntax and not markdown in README file

rallytime commented:
@vutny No worries. It's always better to get this stuff right on the first try! I appreciate the careful review. Is there anything else you see here that should be fixed before I merge this in?

vutny commented on README.rst:
```diff
@@ -251,9 +251,21 @@ Ubuntu and derivatives
-- Ubuntu 10.x/11.x/12.x/13.x/14.x/15.x/16.04
+- Ubuntu 12.04/14.04/16.04
+
+Ubuntu Best Effort Support: Non-LTS releases 
```
Okay, this looks like another sub-section to me, so it can be marked accordingly:
```
Ubuntu Best Effort Support: Non-LTS releases
********************************************
```

Maybe we could add some table of contents later.

vutny reviewed (changes requested):
@rallytime I promise this is a last one request :smile: 
The changes look perfect to me. After the final formatting cleanup will be made this PR could be merged.
Thank you very much!

Commit:
Make sub-section for non-lts ubuntu releases note

rallytime commented:
Alright, I have made the final formatting change, so I will merge this in. Yay!

I created Issue #1008 to track the addition of TOC for the README file. It's good idea.
